procwrap/procwrap.py

```diff
@@ -2,39 +2,157 @@
+            nsqd_http_address=_NSQD_HTTP_ADDRESS):
+
+        self.command = command
+        self.pid_file = pid_file
+        self.disable_nsq = disable_nsq
+        self.nsq_log_topic = nsq_log_topic
+
+        nsqd_http_address = nsqd_http_address.rstrip('/')
+        if not nsqd_http_address.startswith("http://"):
+            nsqd_http_address = 'http://%s' % nsqd_http_address
+
+        self.nsqd_http_address = nsqd_http_address
+
+        self.process = None
+        self.write_to_nsq = lambda msg: None
+
+        if not self.disable_nsq:
+            self.init_nsq()
+
+    def init_nsq(self):
+        session = requests.Session()
+        # TODO use urljoin ?
+        response = None
+        try:
+            response = session.get("%s/ping" % self.nsqd_http_address)
+            if response.status_code != 200:
+                raise Exception("bad response %s" % response)
+        except:
+            logger.exception(
+                "if you wish to disable nsqd, use --disable-nsq"
+            )
+            sys.exit(-1)
+
+        # TODO nsq is a failure point for our logs. what do we do if nsq is down ?
+
+        url = "%s/pub" % self.nsqd_http_address
+        topic = self.nsq_log_topic
+        # TODO what if this write to nsq fails ?
+        self.write_to_nsq = lambda msg: session.post(url, data=msg, params={"topic": topic})
+
+    def start(self):
+        """
+        starts the command and blocks till it finishes.
+        once finished, returns the returncode.
+        """
+
+        if self.pid_file is not None:
+            d = os.path.dirname(self.pid_file)
+            if not os.path.exists(d):
+                os.makedirs(d)
+
+        # NOTE only capturing stderr. stdout still goes to tty.
+
+        self.process = subprocess.Popen(
+            self.command, bufsize=1, universal_newlines=True,
+            stderr=subprocess.PIPE,
+        )
+
+        if self.pid_file is not None:
+            with open(self.pid_file, "w") as fp:
+                fp.write("%s" % self.process.pid)
+
+        read_stderr_line = self.process.stderr.readline
+        poll = self.process.poll
+        # TODO buffering and sending many to nsq together ?
+        # TODO avoid
+        while poll() is None:
+            line = read_stderr_line()
+            # TODO pretty print to stderr ?
+            sys.stderr.write(line)
+
+            # the line may not be valid json object
+            if line.startswith("{") and line.endswith("}\n"):
+                # TODO load the json line, add some other parameters like hostname, uuid, etc
+                # serialize again and send to nsq
+                # TODO check that it is a log
+
+                # NOTE for now just sending it across
+                self.write_to_nsq(line)
+
+            # FIXME is this a bottleneck ? only one log per 0.1 seconds ?
+            # consider threads http://eyalarubas.com/python-subproc-nonblock.html
+            time.sleep(0.1)
```

Comment:
I think would have to add threading or greenlets or tornado for this. I didn't want to do that initially so thought of this approach.

The readline blocks when inactive so its not a problem there.
It is a problem when there are a lot of logs because it will take them one by one with a 0.1 interval.